Tuya: Fix init sequence and handle wifi test command

brandond opened the pull request:
## Description:
* Handle WiFi test command 0x0E
* Run through complete init process as per documentation (heartbeat -> query product -> query conf -> report wifi -> query data packet)
* Handle Product ID string properly, and report in config dump

Documentation ref: https://docs.tuya.com/docDetail?code=K8uhkesnz5mie

## Checklist:
  - [x] The code change is tested and works locally.

brandond committed:
Handle WiFi test command

Also rename commands to match Tuya protocol docs

brandond committed:
Fix init sequence and product info check

OttoWinter commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -8,7 +8,7 @@ namespace tuya {
 
 void Tuya::setup() {
-  this->send_empty_command_(TuyaCommandType::MCU_CONF);
+  this->send_empty_command_(TuyaCommandType::HEARTBEAT);
```
Heartbeat is already sent in the line below (will be executed directly after `setup()`).

Do you know what the MCU_CONF command was for?

Also CC: @ssieb

ssieb replied:
It asks the MCU if the wifi CPU has control of the wifi status led and wifi reset or if the MCU does (usual case).  You can see the config message if we get the response with GPIO numbers.

brandond replied:
Yeah, I know that the heartbeat will get sent in the interval callback but sending it immediately cuts down on the startup time by sending the initial command immediately. I can just wait for the first callback to fire if you think that's cleaner.

ssieb replied:
There shouldn't be any delay, are you seeing something?  The CONF command is sent in startup which triggers the chain.  None of the commands are really required anyway as far as I can tell.

brandond replied:
I just mean, start the request-response chain immediately rather than waiting for the first callback, since we now chain initial heartbeat -> product -> conf as the documentation says you should.

My experience with the Tuya MCUs is that they behave differently depending on version and type. While some will let you skip ahead and just run any command at any time, others are more fussy and you have to follow the documented sequence if you want it to work. Especially with the low-power devices where the MCU can reset or power down the ESP if it doesn't like what it sees.

ssieb replied:
The chain did start immediately, but I suppose it's possible that some device might want the heartbeat first.

OttoWinter replied:
@brandond 
> Yeah, I know that the heartbeat will get sent in the interval callback but sending it immediately cuts down on the startup time by sending the initial command immediately.

Yes, it will make startup time quicker if sent immediately. But `set_interval` *does* execute the lambda immediately after. After that it executes the function at a steady interval with a random offset.

I understand the Tuya protocol a bit better now - the figure in section 2.1 of the page you linked shows we should first send a heartbeat packet. So the change is good 👍

My suggestion: Remove this line as the next line will send it immediately anyway.

brandond replied:
Ah OK. I didn't know that it fired it immediately. Will change.

OttoWinter commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -22,6 +22,7 @@ void Tuya::loop() {
 
 void Tuya::dump_config() {
   ESP_LOGCONFIG(TAG, "Tuya:");
+  ESP_LOGCONFIG(TAG, "  Product: %s", this->product_.c_str());
```
```suggestion
  ESP_LOGCONFIG(TAG, "  Product: '%s'", this->product_.c_str());
```

If the component is not receiving anything, the log output is not ideal

OttoWinter commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -105,39 +106,53 @@ void Tuya::handle_char_(uint8_t c) {
       for (int i = 0; i < len; i++) {
-        if (buffer[i] == 0x00) {
-          valid = true;
+        if (! isprint(buffer[i])) {
```
```suggestion
        if (!std::isprint(buffer[i])) {
```

Whitespace doesn't make travis happy - and std:: just for good measure.

OttoWinter commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -105,39 +106,53 @@ void Tuya::handle_char_(uint8_t c) {
-      this->send_command_(TuyaCommandType::WIFI_STATE, &c, 1);
-      this->send_empty_command_(TuyaCommandType::QUERY_STATE);
+      c[0] = 0x03;
+      c[1] = 0x00;
```
This is unnecessary because only one byte is sent.

Also, I think we should scope `c` differently.

Move `c` into each case where it is used (with curly brace syntax `case ...: {`)

ssieb replied:
Ah, that's how to do it.  I'm not that familiar with C++.

OttoWinter commented on esphome/components/tuya/tuya.h:
```diff
@@ -34,14 +34,24 @@ struct TuyaDatapointListener {
-  SET_DATAPOINT = 0x06,
-  STATE = 0x07,
-  QUERY_STATE = 0x08,
+  DP_DELIVER = 0x06,
```
Hmm, I don't really like the `dp` abbreviation. At least when I read the code initially for this component, it took me a while to understand that dp stood for datapoint. And besides its written out fully everywhere else here too.

```suggestion
  DATAPOINT_DELIVER = 0x06,
```

ssieb replied:
If you're going to duplicate the values like that, might as well separate them into send and receive groups, so it's more clear.  In most cases, the same value is used for the request and the response.

brandond replied:
Tuya uses the DP initialism all over the place (dpid, etc) so I kinda got used to it. I can expand it out. I think keeping them in order of value and not duplicating for rx and tx is fine, since (as you noted) the same command code is valid for both sides of the transaction.

brandond committed:
Fix clang-format suggestions

brandond committed:
Additional changes based on code review

brandond committed:
Fix temp command buffer scope

brandond commented:
FYI I just picked up one of these: https://www.amazon.com/gp/product/B0756CQ5KF/

It's three switches with no dimmer. I am guessing they each present as a separate switch datapoint with no dimmer, which will require change to the light component since it currently requires dimming as well. @OttoWinter @ssieb any preferences on how we handle this?

ssieb commented:
How do you plan to trigger the wifi test command?

ssieb commented:
Yes, I had intended to add an output component as well to handle the boolean datapoints individually.

brandond commented:
@ssieb the wifi test command gets sent by the MCU to the ESP. All you have to do it respond to it once with [0x00, 0x00] and it stops sending it. Mine keeps spamming it once per heartbeat interval otherwise.

ssieb commented:
Oh interesting.  Ok, good to know.  I only have the one device to test with and I bricked two of them. :-)
I wonder why they went with the Tuya MCU system on those.  I have the Sonoff equivalent and they just use the GPIOs directly.  What's the benefit of adding another MCU for something so simple?
Do you have the config output that lists the datapoints?

ssieb commented:
Looking more at the pictures, it looks like there are 2 LEDs for each switch, so that could explain the dedicated MCU.

brandond commented:
Tuya seems to require a MCU even for simple designs that could be done without one. You'd think it would increase the BoM costs but I guess not?

brandond commented:
This thing actually has more going on than I would have suspected. It's not just another prefab esp daughterboard slotted in to a boilerplate reference design, it appears to have been laid out from scratch.

https://imgur.com/a/3vzfOi5

OttoWinter commented:
> It's three switches with no dimmer. I am guessing they each present as a separate switch datapoint with no dimmer, which will require change to the light component since it currently requires dimming as well.

Yeah I think we can change the light component to make the dimmer datapoint optional. I think the logic is mostly compatible - though the validating part could get a little more complicated, maybe we should add a `type` option where the user must set the type of device (binary/dimmer) first.

> Looking more at the pictures, it looks like there are 2 LEDs for each switch, so that could explain the dedicated MCU.

Looking at [this page](https://docs.tuya.com/docDetail?code=K8uhkesn2njgf), it also appears like their main IP is based on the interaction between the MCU and secondary device. This leaves some possibilities:

 - The MCU FW probably existed before the ESP8266.
 - Probably makes cost senses too - with the MCU, Tuya can probably justify their services costing more (because they're doing more)
 - Tuya supports a bunch of WiFi modules: https://docs.tuya.com/docDetail?code=K8uhkoletwzdc - would make sense to keep the HW aspects the same then.
 - For battery powered devices, a secondary MCU for low power mode is obviously a good idea.

brandond commented:
> Yeah I think we can change the light component to make the dimmer datapoint optional. I think the logic is mostly compatible - though the validating part could get a little more complicated, maybe we should add a type option where the user must set the type of device (binary/dimmer) first.

Conceptually, datapoints map most closely to float and binary sensors/outputs, text sensors, etc. The enums are a little weirder. I had planned to just add some new sensor and output platforms that could be used with existing switch, monochromatic, or binary light platforms, but since I didn't ever get around to it I can hardly complain ;)

brandond committed:
Let the interval timer fire the first heatbeat

brandond commented:
Oh, I found another issue with the transmitted commands. It's sending version 0 for everything, which isn't right.

brandond commented:
Changed a few things - realized I was missing a step, and I didn't like how it was reprinting the config after receiving datapoints when we could just wait until init is complete before proceeding. Looks good now:

```
[13:42:18][I][app:100]: ESPHome version 1.15.0-dev compiled on Nov  1 2019, 13:41:51
[13:42:18][C][wifi:409]: WiFi:
[13:42:18][C][wifi:277]:   SSID: 'X'
[13:42:18][C][wifi:278]:   IP Address: 10.0.1.130
[13:42:18][C][wifi:280]:   BSSID: X
[13:42:19][C][wifi:281]:   Hostname: 'dimmer00'
[13:42:19][C][wifi:285]:   Signal strength: -63 dB ▂▄▆█
[13:42:19][C][wifi:289]:   Channel: 1
[13:42:19][C][wifi:290]:   Subnet: 255.255.255.0
[13:42:19][C][wifi:291]:   Gateway: 10.0.1.1
[13:42:19][C][wifi:292]:   DNS1: 10.0.1.1
[13:42:19][C][wifi:293]:   DNS2: (IP unset)
[13:42:19][C][uart:145]: UART Bus:
[13:42:19][C][uart:147]:   TX Pin: GPIO1
[13:42:19][C][uart:150]:   RX Pin: GPIO3
[13:42:19][C][uart:152]:   Baud Rate: 9600 baud
[13:42:19][C][uart:153]:   Stop bits: 1
[13:42:19][C][uart:155]:   Using hardware serial interface.
[13:42:19][C][tuya:023]: Tuya:
[13:42:19][C][tuya:024]:   Product: '{"p":"ezkmcwooh01zaotp","v":"4.0.1","m":1}'
[13:42:19][C][tuya:031]:   Datapoint 3: int value (value: 25)
[13:42:19][C][tuya:029]:   Datapoint 1: switch (value: OFF)
[13:42:19][C][logger:175]: Logger:
[13:42:19][C][logger:176]:   Level: DEBUG
[13:42:19][C][logger:177]:   Log Baud Rate: 0
[13:42:19][C][logger:178]:   Hardware UART: UART1
[13:42:19][C][light:177]: Light 'Dimmer 00 Light'
[13:42:19][C][light:179]:   Default Transition Length: 0.0s
[13:42:19][C][light:180]:   Gamma Correct: 1.00
[13:42:19][C][tuya.light:027]: Tuya Dimmer:
[13:42:19][C][tuya.light:029]:    Dimmer has datapoint ID 3
[13:42:19][C][tuya.light:031]:    Switch has datapoint ID 1
[13:42:19][C][ota:029]: Over-The-Air Updates:
[13:42:19][C][ota:030]:   Address: dimmer00.local:8266
[13:42:19][C][mqtt:051]: MQTT:
[13:42:19][C][mqtt:053]:   Server Address: 10.0.1.27:1883 (10.0.1.27)
[13:42:19][C][mqtt:054]:   Username: ''
[13:42:19][C][mqtt:055]:   Client ID: 'dimmer00-b4e62d7a086b'
[13:42:19][C][mqtt:057]:   Discovery prefix: 'homeassistant'
[13:42:19][C][mqtt:058]:   Discovery retain: YES
[13:42:19][C][mqtt:060]:   Topic Prefix: 'dimmer00'
[13:42:19][C][mqtt:062]:   Log Topic: 'dimmer00/debug'
[13:42:19][C][mqtt:065]:   Availability: 'dimmer00/status'
[13:42:19][C][wifi_signal.sensor:009]: WiFi Signal 'Dimmer 00 WiFi Signal'
[13:42:19][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dB'
[13:42:19][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0
[13:42:19][C][wifi_signal.sensor:009]:   Icon: 'mdi:wifi'
[13:42:19][C][mqtt.light:053]: MQTT Light 'Dimmer 00 Light':
[13:42:19][C][mqtt.light:054]:   State Topic: 'dimmer00/light/dimmer_00_light/state'
[13:42:19][C][mqtt.light:054]:   Command Topic: 'dimmer00/light/dimmer_00_light/command'
[13:42:19][C][mqtt.sensor:024]: MQTT Sensor 'Dimmer 00 WiFi Signal':
[13:42:19][C][mqtt.sensor:028]:   State Topic: 'dimmer00/sensor/dimmer_00_wifi_signal/state'
```

brandond committed:
Fix init steps; add logging

OttoWinter commented:
> Conceptually, datapoints map most closely to float and binary sensors/outputs, text sensors, etc. The enums are a little weirder. I had planned to just add some new sensor and output platforms that could be used with existing switch, monochromatic, or binary light platforms, but since I didn't ever get around to it I can hardly complain ;)

I guess it depends on how modular these devices are.

If these datapoints are used for full RGB or CWWW stuff, then yes we should create them as float outputs. If it's just binary/dimmer lights, then the current design is fine.

OttoWinter commented on esphome/components/tuya/tuya.h:
```diff
@@ -34,19 +34,30 @@ struct TuyaDatapointListener {
 class Tuya : public Component, public uart::UARTDevice {
  public:
   float get_setup_priority() const override { return setup_priority::HARDWARE; }
+  bool can_proceed() override { return this->init_state_ == TuyaInitState::INIT_DONE; }
```
Hmm, this thing has setup priority HARDWARE (so before WiFi). If the tuya component blocks, wifi will never start. Plus wifi setup will be delayed until tuya setup is complete, which could be bad for low-power devices.

What's the reason for `can_proceed` ? I vaguely remember something about MQTT setup back when you first started with the tuya PIR sensor, but don't know the specifics.

brandond replied:
Good point. It probably does not need to be HARDWARE, it can go after wifi and probably should since we're telling the MCU that wifi and mqtt are both up in the wifi status step.

I added the can_proceed so that we don't dump config or set up lights until after the init process is complete. Previously dump_config would get called before init was complete and we didn't have product info or datapoints yet. This was worked around by calling dump_config again whenever new datapoints are discovered, which seemed wrong - especially given the warning you added when there are no datapoints.

The tuya protocol is designed to not proceed past the wifi config messages until everything is up. It has various values that can be used to indicate the connection status (disconnected/wifi/wifi+mqtt) and will flash different patterns to indicate each state. It also has status values for smartconfig or AP mode that I used in my low-power Tuya component to indicate OTA in progress. We just jump in at the end of the sequence and tell it that everything is up, which is fine, but it means that there's less feedback about what's going on while the device is booting. I could fix that but it seemed like a low priority given that most wired Tuya devices don't reboot as often as the battery-powered ones do.

OttoWinter replied:
Now that I think of it - there may be another issue here.

If we set setup priority `HARDWARE` and the Tuya component doesn't work (for whatever reason), the device will not connect to wifi -> bad
If we set it to something later (like `AFTER_WIFI` or `LATE`), the tuya component won't be set up until after WiFi is connected. So any automations that run before that will fail -> also not good.

So I think we should remove `can_proceed`; I understand that it makes the datapoint dumping better, but the disadvantages are not that great.

> Previously dump_config would get called before init was complete and we didn't have product info or datapoints yet.

Does the init process take that long? I mean by the time dump_config gets called through the "normal" way the device is already fully setup (meaning connected to wifi, which can take a few secs).

> It also has status values for smartconfig or AP mode that I used in my low-power Tuya component to indicate OTA in progress.

Just FYI, there's also a global variable that stores the warning/error state of the device - the tuya component could access that to check the state of the device: https://github.com/esphome/esphome/blob/6682c43dfaeb1c006943ae546145e5f22262cadb/esphome/components/status_led/status_led.cpp#L23
But let's leave that for another PR 👍

OttoWinter committed:
Lint

OttoWinter commented:
@brandond I fixed the lint issue - let me know if you think the code is ready know. Also: this is a fix, right? Should I cherry pick it into a patch release?

brandond committed:
Remove setup_priority override

brandond commented:
Without the can_proceed block, the dump occurs before the init sequence is complete. I think this is because the serial port is only polled once per loop, and the loop doesn't run enough times before boot is complete and dump_config gets called.

I agree that this is a display issue, but it'd be nice to have it work properly. Can you think of some way to have it wait a little longer before dumping config, but without blocking boot if something goes wrong? I suppose I could just have the dump_config call set a flag that does the actual dump later when the init completes... or is that too ugly?

OttoWinter commented:
I believe it is already doing that. See the set_timeout call when a datapoint is received. Each time a datapoint is received, a timeout is set to dump the config after some time. If another datapoint is received in the meantime, the timeout is restarted.

We could do two things to improve that:

 - Use the terminating message after receiving the datapoints for scheduling the dump config.
 - I dump_config, check if end message has been received yet. If not, don't print "no datapoints received".

ssieb commented:
There is no terminating message.  That's why there is a timeout.

OttoWinter commented:
@ssieb Correct - I thought there was one added in this PR at first because `can_proceed` was waiting for *something*.

But now I'm more confused... Yes *a* dump may happen before the full init sequence is complete, but a second dump *will* happen after the full init sequence is complete.

brandond commented:
I removed the self-triggered dump on datapoint discovery when I added an explicit init sequence and can_proceed. It was confusing how it dumped twice, once with a warning about no datapoints and then again later with datapoints.

I think the current state is OK. It seems like it occasionally dumps too early on reboot, but it works reliably on powerup, probably due to a difference in state on the MCU side. The only real loss is the product data string.

ssieb commented:
You only saw that initial config dump because you're connected on serial.  Most people will only be able to connect over wifi and by that time it will be fully configured.

brandond commented:
I can't be connected over serial, serial goes to the MCU ;) All my testing has been via OTA with MQTT logs.

ssieb commented:
I always use the OTA for logs, maybe mqtt is quicker.
Anyway, I've tested it and it still works the same for me, so that's good.

ssieb commented:
Now I switched the logging to verbose and the dimmer is bricked.  It gets an IP address but nothing can connect.

brandond commented:
Smart thing happened to me once, but I just chalked it up to the infamous #455. I can see if I can replicate it again on purpose.

OttoWinter commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -214,8 +238,6 @@ void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
   if (!found) {
     this->datapoints_.push_back(datapoint);
-    // New datapoint found, reprint dump_config after a delay.
-    this->set_timeout("datapoint_dump", 100, [this] { this->dump_config(); });
```
Now that `can_proceed` is away again, can we re-add this line?

Without it, this line will definitely not work for serial-port users - for MQTT/api log users it will depend on how fast the init sequence is.

With this line, the user would at least know that a new datapoint got added.

brandond replied:
I just changed the dump_config override to call the actual dump a few seconds later. This should allow it time to complete the init sequence before dumping.

I didn't like the previous approach of dumping whenever a datapoint was discovered as it would trigger even if debug and/or debug-level logging were not requested.

OttoWinter replied:
I think it's worse that way - now the user needs to wait a whole 2s each time they connect to the device. This is true even when all datapoints have been found a long time ago.

> I didn't like the previous approach of dumping whenever a datapoint was discovered as it would trigger even if debug and/or debug-level logging were not requested.

If that's true that's a bug - debug-level logging has the same level as config-level, just a different color. In fact, dump_config is called regardless of debug level: 

https://github.com/esphome/esphome/blob/a6faccb4d9fee7bdf92fb665247e6c73815c5f66/esphome/core/application.cpp#L103

If the logger level is info or below, then the dump_config calls will just be empty, and not print anything.

> This should allow it time to complete the init sequence before dumping.

I think we can solve that in a better way than having the whole config dump be delayed 2 seconds - just print something if we know we're not ready reading datapoints yet, and re-print once we have received all of them:

```diff
﻿﻿diff --git a/esphome/components/tuya/tuya.cpp b/esphome/components/tuya/tuya.cpp
index 23fcd15a..e16b32d4 100644
--- a/esphome/components/tuya/tuya.cpp
+++ b/esphome/components/tuya/tuya.cpp
@@ -37,6 +37,9 @@ void Tuya::dump_config() {
       ESP_LOGCONFIG(TAG, "  Datapoint %d: unknown", info.id);
   }
   if (this->datapoints_.empty()) {
+    if (this->init_state_ <= TuyaInitState::INIT_DATAPOINT) {
+      ESP_LOGCONFIG(TAG, "  Tuya setup not complete yet - datapoints will be reported when ready.")
+    }
     ESP_LOGCONFIG(TAG, "  Received no datapoints! Please make sure this is a supported Tuya device.");
   }
   this->check_uart_settings(9600);
@@ -239,7 +242,7 @@ void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
   if (!found) {
     this->datapoints_.push_back(datapoint);
     // New datapoint found, reprint dump_config after a delay.
-    this->set_timeout("datapoint_dump", 250, [this] { this->dump_config(); });
+    this->set_timeout("datapoint_dump", 1000, [this] { this->dump_config(); });
   }
 
   // Run through listeners
```

brandond replied:
That works.

> now the user needs to wait a whole 2s each time they connect to the device

I thought that dump_config was just called once when all the components have completed setup. Does the native API behave differently to accommodate sending debug info after the client connects?

OttoWinter replied:
Ah yes, sorry should have mentioned that.

Yes, native API re-schedules all components to dump their config when a client asks for it.

You're correct that with MQTT it's different - there it's only printed once everything is completely set up.

brandond replied:
I changed it a bit. It still triggers config dump when we get datapoints, but the message about supported devices is handled differently due to the fact that init will never move past the initial state on unsupported devices.

brandond committed:
Add delay to dump_config

brandond committed:
Refactor dump sequence

brandond committed:
Fix verbose logging

OttoWinter commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -22,8 +21,12 @@ void Tuya::loop() {
-  if ((gpio_status_ != -1) || (gpio_reset_ != -1))
-    ESP_LOGCONFIG(TAG, "  GPIO MCU configuration not supported!");
+  if (this->init_state_ != TuyaInitState::INIT_DONE) {
+    ESP_LOGCONFIG(TAG, "  Configuration will be reported when setup is complete. Current init_state: %d",
```
```suggestion
    ESP_LOGCONFIG(TAG, "  Configuration will be reported when setup is complete. Current init_state: %u",
```

brandond committed:
Fix lints

brandond commented:
@OttoWinter clang-tidy keeps insisting I use %hhu in the log string, but this is coming out in the logs as literal 'hu'.
```c++
  ESP_LOGV(TAG, "Sending Tuya: CMD=0x%02X VERSION=%u DATA=[%s] INIT_STATE=%hhu", command, version,
           hexencode(buffer, len).c_str(), this->init_state_);

```

```
[15:44:42][V][tuya:266]: Sending Tuya: CMD=0x00 VERSION=0 DATA=[(0)] INIT_STATE=hu
[15:44:42][V][tuya:103]: Received Tuya: CMD=0x00 VERSION=3 DATA=[01 (1)] INIT_STATE=hu
```

brandond committed:
Don't bother suppressing duplicate config dumps

OttoWinter commented:
@brandond Ugh, that's no good. %hhu appears to be a weird printf-extension, and the printf implementation doesn't support it.

We can't disable the clang-tidy check because that would disable all printf-checks, but those are important checks. Best is to add `// NOLINT` to all offending lines.

OttoWinter committed:
nolint
